### PR TITLE
Add code development quick-start guide

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Add a `devcontainer.json` configuration and a quick-start tutorial that allows new students/contributors to jump right into a working development environment using our Docker container and the Visual Studio Code editor.

The devcontainer configuration will also be useful when [GitHub Codespaces](https://github.com/features/codespaces) launch.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
@geoffrey4444 this PR picks up where your #1884 left off. I tried to keep it as concise as possible to quick-start the reader, not giving too many details and referring to existing documentation wherever possible. We can add more "tips and tricks" at the bottom in the future.

**Comment by geoffrey4444** [comment]
Thanks!! I will try to review this week. Looking forward to codespaces, still waiting for my beta invite :(

File: docs/DevGuide/QuickStart.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
+\cond NEVER
+Distributed under the MIT License.
+See LICENSE.txt for details.
+\endcond
+# Code development quick-start {#dev_guide_quick_start}
```
Comment (nilsdeppe) [inline docs/DevGuide/QuickStart.md]: I don't think this should be the "quickstart" guide. We should not be imposing a specific IDE or development workflow...

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (nilsdeppe) [inline .devcontainer/devcontainer.json]: We don't have special environments for any other IDE... I don't like the idea of having a bunch of configs in the repo for IDEs. The only "special" treatment that we've given them is adding their custom files to 'gitignore', which I think is reasonable in order to aid development workflows. More than that is not okay by me.

File: docs/Installation/Installation.md [hunk 3]
```diff
@@ -8,6 +8,11 @@ This page details the installation procedure for SpECTRE on personal computers.
 For instructions on installing SpECTRE on clusters please refer to the \ref
 installation_on_clusters "Installation on clusters" page.
 
+### Quick-start guide for code development
```
Comment (nilsdeppe) [inline docs/Installation/Installation.md]: I don't think this should be here. VSCode is not an editor we are endorsing or anything like that. I'm fine with a `Dev Guide` entry on "Building SpECTRE with VSCode and Docker", but it should **not** make it sound like VSCode or that work flow is at all the recommended workflow.

File: containers/buildenv.vscode/cmake-kits.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+[
```
Comment (nilsdeppe) [inline containers/buildenv.vscode/cmake-kits.json]: Again, I don't think it's reasonable to start adding a bunch of files to special treat editors. You probably don't want me to dump my entire Emacs config in ;)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: docs/DevGuide/QuickStart.md (hunk 1 above)
Comment (geoffrey4444) [reply docs/DevGuide/QuickStart.md]: How about @quick start with docker & visual studio code" along with language that this is a quick start, not the quick start?

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (geoffrey4444) [reply .devcontainer/devcontainer.json]: How about putting them in another repo then? Surely they should be shared somehow

File: docs/Installation/Installation.md (hunk 3 above)
Comment (geoffrey4444) [reply docs/Installation/Installation.md]: I agree in general...no one is telling you or others to switch! But it is an especially easy way to get started for new developers, so I think it should be ok to say it's something new devs might want to consider, especially if they aren't experienced setting up their own dev environment, along with language that they can use tools of their choice if they wish and that we don't endorse or preferentially support any particular choice

File: containers/buildenv.vscode/cmake-kits.json (hunk 4 above)
Comment (geoffrey4444) [reply containers/buildenv.vscode/cmake-kits.json]: Maybe not in the spectre repo, though I think a case could be made for support/DevEnvExamples or something like that, but some repo makes sense! You've emailed your enacts file on request, why not keep it around as an example for others?

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (nilsvu) [reply .devcontainer/devcontainer.json]: I agree we shouldn't endorse a particular editor or something like that. I do suggest to keep this `devcontainer.json` in the repository however, to support GitHub's [codespaces](https://docs.github.com/en/github/developing-online-with-codespaces/configuring-codespaces-for-your-project) once they launch. We can wait with adding the `devcontainer.json` to the repository until the GitHub feature has actually launched if people prefer. Here's the docs:

- https://docs.github.com/en/github/developing-online-with-codespaces/configuring-codespaces-for-your-project

File: containers/buildenv.vscode/cmake-kits.json (hunk 4 above)
Comment (nilsdeppe) [reply containers/buildenv.vscode/cmake-kits.json]: I do keep my emacs config around for others to use if interested: https://github.com/nilsdeppe/MyEnvironment (though the main reason is selfish: I can't lose the configs if they're in a repo ;) ).

I'm not against sharing the environment, I think `support/DevEnvExamples` is a reasonable place to put configs. I know SpEC has its Emacs input file syntax highlighting script somewhere, and I think that's a good thing. I'm not sure if the `json` files need to be in a specific location. If they do I think it's reasonable to have in the instructions "Copy file X to Y using `cp ...`". We can add where the file needs to get copied to the `.gitignore` list. Does that sound okay?

File: docs/DevGuide/QuickStart.md (hunk 1 above)
Comment (nilsdeppe) [reply docs/DevGuide/QuickStart.md]: That's fine by me

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (nilsdeppe) [reply .devcontainer/devcontainer.json]: I'm rather hesitant on supporting features that don't yet exist (or are in beta). Can we only add the files into `support/DevEnvExamples` or `support/DevEnvs` needed for local VSCode with instructions on where to put them? We can `gitignore` the locations they need to go.

File: containers/buildenv.vscode/cmake-kits.json (hunk 4 above)
Comment (geoffrey4444) [reply containers/buildenv.vscode/cmake-kits.json]: Sounds perfect to me! We'll just add a readme saying "copy .vscode to $SPECTRE_HOME"inside support/DevEnvExamples/vscode

Would be nice if separately you put your example env in support/DevEnvExamples/emacs or whatever you want to name it, so it's easier to find, if you're willing to

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (nilsvu) [reply .devcontainer/devcontainer.json]: Fine with me. Once GitHub launches the feature it will be easy enough to transition.

**Comment by nilsvu** [comment]
Fixups posted @nilsdeppe @geoffrey4444!

**Comment by nilsvu** [comment]
@nilsdeppe @geoffrey4444 please take a look at the fixups

**Review by nilsdeppe — COMMENTED** [review]
LGTM. Please go ahead and squash :) Sorry that this got forgotten/lost

**Review by nilsdeppe — APPROVED** [review]

**Review by geoffrey4444 — APPROVED** [review]

**Pull request merged**